File: src/main/java/com/gmail/samueler53/fwhumanstratego/listeners/onPlayerTeleportListener.java
package com.gmail.samueler53.fwhumanstratego.listeners;

import com.gmail.samueler53.fwhumanstratego.FwHumanStratego;
import com.gmail.samueler53.fwhumanstratego.gui.TeamGui;
import com.gmail.samueler53.fwhumanstratego.managers.GameManager;
import com.gmail.samueler53.fwhumanstratego.message.Message;
import com.gmail.samueler53.fwhumanstratego.objects.Game;
import com.gmail.samueler53.fwhumanstratego.objects.Squad;
import org.bukkit.Bukkit;
import org.bukkit.GameMode;
import org.bukkit.entity.Player;
import org.bukkit.event.EventHandler;
import org.bukkit.event.Listener;
import org.bukkit.event.player.PlayerTeleportEvent;
import org.bukkit.potion.PotionEffect;
import org.bukkit.potion.PotionEffectType;

import java.util.UUID;

public class onPlayerTeleportListener implements Listener {

    GameManager gameManager = GameManager.getInstance();

    @EventHandler
    public void onPlayerTeleport(PlayerTeleportEvent event) {
        Player player = event.getPlayer();
        UUID uuid = player.getUniqueId();
        if (gameManager.isPlayerPlaying(uuid).isPresent()) {
            Game game = gameManager.getGameWherePlayerPlaying(uuid);
            if (game.getArena().getLobbyLocation().equals(event.getTo())) {
                lobbyTeleport(uuid, game);
            } else if (game.getArena().getRedTeamLocation().equals(event.getTo()) || game.getArena().getBlueTeamLocation().equals(event.getTo())) {
                spawnTeleport(uuid, game);
            }
        }
    }

    private void lobbyTeleport(UUID uuid, Game game) {
        Player player = Bukkit.getPlayer(uuid);
        assert player != null;
        Message.GAME_CHOOSETEAM.send(player);
        TeamGui teamGui = game.getTeamGui();
        Bukkit.getScheduler().runTaskLater(FwHumanStratego.getPlugin(), () -> {
            teamGui.show(uuid);
        }, 1L);
    }

    private void spawnTeleport(UUID uuid, Game game) {
        Player player = Bukkit.getPlayer(uuid);
        assert player != null;
        Message.GAME_CHOOSEROLE.send(player);
        Squad squad = game.getSquadFromPlayer(uuid);
        Bukkit.getScheduler().runTaskLater(FwHumanStratego.getPlugin(), () -> {
            squad.getRoleGui().show(uuid);
        }, 1L);
        if (game.isSpectateMode(squad)) {
            player.setGameMode(GameMode.SPECTATOR);
        }
        selectionRoleEffects(uuid);
    }

    private void selectionRoleEffects(UUID uuid) {
        Player player = Bukkit.getPlayer(uuid);
        assert player != null;
        player.addPotionEffect(new PotionEffect(PotionEffectType.JUMP, 2147483647, 250));
        player.setWalkSpeed(0.0F);
        player.setFoodLevel(4);
    }
}

File: src/main/java/com/gmail/samueler53/fwhumanstratego/listeners/OnPlayerTeleportListener.kt
package com.gmail.samueler53.fwhumanstratego.listeners

import com.gmail.samueler53.fwhumanstratego.managers.GameManager
import com.gmail.samueler53.fwhumanstratego.message.Message
import com.gmail.samueler53.fwhumanstratego.objects.Game
import com.gmail.samueler53.fwhumanstratego.utils.launch
import kotlinx.coroutines.delay
import org.bukkit.GameMode
import org.bukkit.entity.Player
import org.bukkit.event.EventHandler
import org.bukkit.event.Listener
import org.bukkit.event.player.PlayerTeleportEvent
import org.bukkit.potion.PotionEffect
import org.bukkit.potion.PotionEffectType

class OnPlayerTeleportListener : Listener {

    @EventHandler
    fun onPlayerTeleport(event: PlayerTeleportEvent) {
        val game = GameManager.getGameForPlayer(event.player) ?: return
        if (game.arena.lobbyLocation == event.to) {
            lobbyTeleport(event.player, game)
            return
        }
        if (game.arena.redTeamLocation == event.to || game.arena.blueTeamLocation == event.to) {
            spawnTeleport(event.player, game)
        }
    }

    private fun lobbyTeleport(player: Player, game: Game) {
        Message.GAME_CHOOSETEAM.send(player)
        val teamGui = game.teamGui
        launch {
            delay(1L)
            teamGui.show(player)
        }
    }

    private fun spawnTeleport(player: Player, game: Game) {
        Message.GAME_CHOOSEROLE.send(player)
        val team = game.getTeamForPlayer(player)
        launch {
            delay(1L)
            team.roleGui.show(player)
        }
        if (game.isSpectateMode(team)) {
            player.gameMode = GameMode.SPECTATOR
        }
        player.addPotionEffect(PotionEffect(
            PotionEffectType.JUMP,
            2147483647,
            250
        ))
        player.walkSpeed = 0.0f
        player.foodLevel = 4
    }

}